Added the String Compression Algorithm #3911

Swarga-codes opened the pull request:
<!-- For completed items, change [ ] to [x] -->

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

Swarga-codes commented:
Should I remove the main function from the code and make a test file in the test folder for this particular algorithm?

Swarga-codes commented:
@siriak  I made the necessary changes as u told on the PR, the changes include:

Reduce the size of code by dividing the code into smaller functions
Removed unnecessary comments
Removed the main
Included some Junit test cases for the code which you can access in the tests folder
Can you please check ?

siriak commented on src/main/java/com/thealgorithms/strings/StringCompression.java:
```diff
@@ -0,0 +1,65 @@
+   * @param i     the index at which the count is to be appended
+   * @return the res string appended with the count
+   */
+  public static String checkCount(String res, int count, char[] ch, int i) {
```
This funciton only uses `ch[i]`, so you could only pass the current character to it and not the whole array

siriak commented on src/main/java/com/thealgorithms/strings/StringCompression.java:
```diff
@@ -0,0 +1,65 @@
+    // Keeping the count as 1 since every element present will have atleast a count
+    // of 1
+    int count = 1;
+    String res = "";
```
```suggestion
    String compressedString = "";
```

siriak commented on src/main/java/com/thealgorithms/strings/StringCompression.java:
```diff
@@ -0,0 +1,65 @@
+   * @param ch character array that contains the group of characters to be encoded
+   * @return the compressed character array as string
+   */
+  public static String compress(char[] ch) {
```
```suggestion
  public static String compress(String input) {
```

siriak commented on src/main/java/com/thealgorithms/strings/StringCompression.java:
```diff
@@ -0,0 +1,65 @@
+package com.thealgorithms.strings;
+/*  Link to the problem : https://leetcode.com/problems/string-compression/description/
```
Please link to Wikipedia instead

siriak commented on src/main/java/com/thealgorithms/strings/StringCompression.java:
```diff
@@ -0,0 +1,65 @@
+        }
+      }
+    }
+    // This loop is not necessary unless you are doing it for leetcode submission
+    for (int i = 0; i < res.length(); i++) {
+      ch[i] = res.charAt(i);
+    }
```
```suggestion
```

siriak commented on src/main/java/com/thealgorithms/strings/StringCompression.java:
```diff
@@ -0,0 +1,65 @@
+        count = count + 1;
+      }
+      if ((i + 1) == ch.length - 1 && ch[i + 1] == ch[i]) {
+        res = checkCount(res, count, ch, i);
```
Please provide a more descriptive name instead of `checkCount` because I'd expect `check...` function to throw an error and to not return a string

siriak commented on src/test/java/com/thealgorithms/strings/StringCompressionTest.java:
```diff
@@ -0,0 +1,24 @@
+
+public class StringCompressionTest {
+    @Test
+    public void StringCompressionTest(){
```
Could you make it a parameterized test?

Swarga-codes commented:
I made the necessary changes as per your instructions, that is:
1) Changed the parameters for the checkCount function and passed the particular character rather than passing the entire array.
2) Changed the name of the resulting string "res" to "compressedString".
3) Changed the type of input from a character array to a string named "input".
4) Changed the reference link from Leetcode to Wikipedia.
5) Removed the unnecessary loop at the end as suggested.
6) Changed the name of checkCount function to a more descriptive name, that is, appendCount, since we are appending the count to the resulting string.
7) Changed the Junit test cases to parameterized Junit tests.
Can you please check?

siriak reviewed (approved):
Looks good, thanks!
